### PR TITLE
Add send_signal method to pebble.Client and model.Container

### PR DESCRIPTION
**Pull request by benhoyt**
This is the Python Operator Framework client part of the send-signals spec. The new `/v1/signals` Pebble API is being added in https://github.com/canonical/pebble/pull/92

### CODE REVIEW

**Review by jnsgruk — APPROVED** [review]
Nice, thanks Ben!

File: ops/pebble.py [hunk 1]
```diff
@@ -1926,3 +1926,30 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
         base_url = self.base_url.replace('http://', 'ws://')
         url = '{}/v1/tasks/{}/websocket/{}'.format(base_url, task_id, websocket_id)
         return url
+
+    def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        """Send the given signal to the list of services named.
+
+        Args:
+            sig: Name or number of signal to send, e.g., "SIGHUP", 1, or
+                signal.SIGHUP.
```
Comment (hpidcock) [inline ops/pebble.py:1935]: is signal.SIGHUP a string or int? Otherwise should it be in the union typing?

File: ops/pebble.py [hunk 2]
```diff
@@ -1926,3 +1926,30 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
...
+        """
+        if not isinstance(services, (list, tuple)):
+            raise TypeError('services must be a list of str, not {}'.format(
+                type(services).__name__))
+        for s in services:
+            if not isinstance(s, str):
+                raise TypeError('service names must be str, not {}'.format(type(s).__name__))
+
+        if isinstance(sig, int):
+            sig = signal.Signals(sig).name
```
Comment (hpidcock) [inline ops/pebble.py:1950]: why is this .name when if someone passes a predefined signal from the signal module it is just set on the body?

File: test/test_model.py [hunk 3]
```diff
@@ -1199,6 +1199,22 @@ def test_exec(self):
...
+        with self.assertRaises(TypeError):
+            self.container.send_signal('SIGHUP')
+
+        self.container.send_signal('SIGHUP', 's1')
+        self.assertEqual(self.pebble.requests, [
+            ('send_signal', 'SIGHUP', ('s1',)),
+        ])
+        self.pebble.requests = []
+
+        self.container.send_signal('SIGHUP', 's1', 's2')
```
Comment (hpidcock) [inline test/test_model.py:1212]: add test for the signal.SIGHUP case?

**Review by hpidcock — APPROVED** [review]
Couple of comments, but seems good

File: ops/pebble.py (hunk 1 above)
Comment (benhoyt) [reply ops/pebble.py:1935]: Yeah, it's a bit weird. `signal.SIGxyz` are defined as an `IntEnum` which basically means the values are an `int` subclass, kind of like `type Signal int` in Go.

```python
>>> isinstance(signal.SIGHUP, int)
True
```

File: ops/pebble.py (hunk 2 above)
Comment (benhoyt) [reply ops/pebble.py:1950]: See above, the `isinstance` check will return `True` for `signal.SIGHUP` and friends. Arguably a separate `isinstance` check for `signal.Signals` would be more explicit, but it's unnecessary.

File: test/test_model.py (hunk 3 above)
Comment (benhoyt) [reply test/test_model.py:1212]: I don't think it's necessary given that `model.Container.send_signal` is basically a one-line wrapper around `self._pebble.send_signal(sig, service_names)` (which does test this).

**Pull request merged**